G1ANT.Addon.Images/G1ANT.Addon.Images/Api/Imaging.cs

```diff
@@ -46,97 +51,114 @@ public static bool AreImagesEqual(Bitmap image1, Bitmap image2)
-                                    break;
-                                }
+            if (resultCollection.Count() > 0)
+            {
+                var rectangle = resultCollection.FirstOrDefault();
+                return new Rectangle(rectangle.X, rectangle.Y, needle.Width, needle.Height);
+            }
+            else return Rectangle.Empty;
+        }
 
-                                pBig++;
-                                pSmall++;
-                            }
+        private static int[][] GetPixelArray(Bitmap bitmap)
+        {
+            var result = new int[bitmap.Height][];
+            var bitmapData = bitmap.LockBits(new Rectangle(0, 0, bitmap.Width, bitmap.Height), ImageLockMode.ReadOnly,
+                PixelFormat.Format32bppArgb);
 
-                            if (!matchFound) break;
+            for (int y = 0; y < bitmap.Height; ++y)
+            {
+                result[y] = new int[bitmap.Width];
+                Marshal.Copy(bitmapData.Scan0 + y * bitmapData.Stride, result[y], 0, result[y].Length);
+            }
 
-                            pSmall = pSmallBackup;
-                            pBig = pBigBackup;
+            bitmap.UnlockBits(bitmapData);
 
-                            pSmall += smallStride * (1 + i);
-                            pBig += bigStride * (1 + i);
-                        }
+            return result;
+        }
 
-                        if (matchFound)
-                        {
-                            location.X = x;
-                            location.Y = y;
-                            location.Width = smallBmp.Width;
-                            location.Height = smallBmp.Height;
-                            break;
-                        }
-                        else
-                        {
-                            pBig = pBigBackup;
-                            pSmall = pSmallBackup;
-                            pBig += 3;
-                        }
+        private static IEnumerable<Point> FindMatch(IEnumerable<int[]> haystackLines, int[] needleLine, double tolerance)
+        {
+            var y = 0;
+            foreach (var haystackLine in haystackLines)
+            {
+                for (int x = 0, n = haystackLine.Length - needleLine.Length; x < n; ++x)
+                {
+                    if (ContainSameElements(haystackLine, x, needleLine, 0, needleLine.Length, tolerance))
+                    {
+                        yield return new Point(x, y);
                     }
-
-                    if (matchFound) break;
-
-                    pBig += bigOffset;
                 }
+                y += 1;
             }
+        }
 
-            bigBmp.UnlockBits(bigData);
-            smallBmp.UnlockBits(smallData);
+        private static bool ContainSameElements(int[] first, int firstStart, int[] second, int secondStart, int length, double tolerance)
+        {
+            if (tolerance > 0)
```

Comment:
I think this whole function can be reduced to:

        private static bool ContainSameElements(int[] first, int firstStart, int[] second, int secondStart, int length, double tolerance)

        {
            return !second
                .Where((t, i) =>
                first[i + firstStart] - tolerance >= second[i + secondStart] ||
                first[i + firstStart] + tolerance <= second[i + secondStart])
                .Any();
        }

Reply:
tolerance is not included into logic

Reply:
It is included, but as we talked, your solution is faster